File: node/token_throttle.js
/**
 * token_throttle - Simple token rate-limit management for LLM API calls.
 *
 * Zero dependencies. Drop into any project.
 *
 * Usage:
 *   const { TokenThrottle } = require('./token_throttle');
 *   const throttle = new TokenThrottle({ tokensPerMinute: 30000 });
 *
 *   for await (const chunk of throttle.consume(largeText)) {
 *     const response = await callYourApi(chunk);
 *   }
 *
 * Sub-agent usage (enforces shared budget automatically):
 *   const { SubAgentThrottle } = require('./token_throttle');
 *   const throttle = new SubAgentThrottle(); // caps to 15K TPM by default
 *
 * Strict mode (throws BudgetExceeded instead of waiting):
 *   const { TokenThrottle, BudgetExceeded } = require('./token_throttle');
 *   const throttle = new TokenThrottle({ budget: 15000, strict: true });
 *   try {
 *     await throttle.waitIfNeeded(25000);
 *   } catch (e) {
 *     if (e instanceof BudgetExceeded) { ... }
 *   }
 */

class TokenThrottle {
  /**
   * @param {Object} options
   * @param {number} [options.tokensPerMinute=30000] - Your TPM limit.
   * @param {number} [options.margin=0.85] - Safety factor (0-1).
   * @param {number|null} [options.budget=null] - Hard cap on effective TPM. Ceiling =
   *   min(budget, tpm * margin). Useful for sub-agents sharing an API key.
   * @param {number} [options.chunkSize] - Max tokens per chunk. Defaults to budget.
   * @param {number} [options.charsPerToken=4] - Characters-per-token ratio.
   * @param {boolean} [options.strict=false] - Throw BudgetExceeded instead of waiting
   *   when a single request exceeds the budget.
   */
  constructor({
    tokensPerMinute = 30000,
    margin = 0.85,
    budget = null,
    chunkSize = null,
    charsPerToken = 4,
    strict = false,
  } = {}) {
    this.tpm = tokensPerMinute;
    this.margin = margin;
    this.strict = strict;
    const effective = Math.floor(tokensPerMinute * margin);
    this.budget = budget !== null ? Math.min(budget, effective) : effective;
    this.chunkSize = chunkSize || this.budget;
    this.charsPerToken = charsPerToken;

    this._tokensUsed = 0;
    this._windowStart = Date.now();
  }

  // ── Estimation ──────────────────────────────────────────────

  /**
   * Estimate token count from text length.
   * @param {string} text
   * @returns {number}
   */
  estimate(text) {
    return Math.max(1, Math.floor(text.length / this.charsPerToken));
  }

  // ── Window management ───────────────────────────────────────

  _maybeResetWindow() {
    const elapsed = Date.now() - this._windowStart;
    if (elapsed >= 25000) {
      this._tokensUsed = 0;
      this._windowStart = Date.now();
    }
  }

  reset() {
    this._tokensUsed = 0;
    this._windowStart = Date.now();
  }

  // ── Throttling ──────────────────────────────────────────────

  /**
   * Reserve tokens from the budget, sleeping if necessary.
   * @param {number} tokens
   * @returns {Promise<number>} Milliseconds slept.
   * @throws {BudgetExceeded} If strict=true and tokens exceeds the budget.
   */
  async waitIfNeeded(tokens) {
    if (this.strict && tokens > this.budget) {
      throw new BudgetExceeded(
        `Request of ${tokens} tokens exceeds budget of ${this.budget} TPM. ` +
        "Reduce the request size or disable strict mode."
      );
    }

    this._maybeResetWindow();

    if (this._tokensUsed + tokens <= this.budget) {
      this._tokensUsed += tokens;
      return 0;
    }

    const elapsed = Date.now() - this._windowStart;
    const sleepMs = Math.max(0, 25000 - elapsed + 500); // +500ms buffer

    if (sleepMs > 0) {
      await new Promise((resolve) => setTimeout(resolve, sleepMs));
    }

    this._tokensUsed = tokens;
    this._windowStart = Date.now();
    return sleepMs;
  }

  // ── Chunking ────────────────────────────────────────────────

  /**
   * Split text into chunks of roughly maxChars, respecting boundaries.
   * @param {string} text
   * @param {number} maxChars
   * @returns {string[]}
   */
  _splitText(text, maxChars) {
    if (text.length <= maxChars) return [text];

    const chunks = [];
    let remaining = text;

    while (remaining.length > 0) {
      if (remaining.length <= maxChars) {
        chunks.push(remaining);
        break;
      }

      const slice = remaining.slice(0, maxChars);
      let breakPos;

      // Try paragraph boundary
      breakPos = slice.lastIndexOf("\n\n");

      // Fall back to sentence boundary
      if (breakPos < maxChars / 4) {
        const sentenceMatch = [...slice.matchAll(/[.!?]\s+/g)];
        const last = sentenceMatch[sentenceMatch.length - 1];
        breakPos =
          last && last.index + last[0].length > maxChars / 4
            ? last.index + last[0].length
            : -1;
      }

      // Fall back to word boundary
      if (breakPos < maxChars / 4) {
        breakPos = slice.lastIndexOf(" ");
      }

      // Hard cut
      if (breakPos < maxChars / 4) {
        breakPos = maxChars;
      }

      chunks.push(remaining.slice(0, breakPos).trimEnd());
      remaining = remaining.slice(breakPos).trimStart();
    }

    return chunks;
  }

  // ── Main interface ──────────────────────────────────────────

  /**
   * Async generator that yields throttled chunks of text.
   * @param {string} text
   * @yields {string}
   */
  async *consume(text) {
    const maxChars = Math.floor(this.chunkSize * this.charsPerToken);
    const chunks = this._splitText(text, maxChars);

    for (const chunk of chunks) {
      const tokens = this.estimate(chunk);
      await this.waitIfNeeded(tokens);
      yield chunk;
    }
  }

  // ── Utilities ───────────────────────────────────────────────

  get remainingTokens() {
    this._maybeResetWindow();
    return Math.max(0, this.budget - this._tokensUsed);
  }

  get secondsUntilReset() {
    const elapsed = Date.now() - this._windowStart;
    return Math.max(0, (25000 - elapsed) / 1000);
  }

  toString() {
    return `TokenThrottle(tpm=${this.tpm}, budget=${this.budget}, used=${this._tokensUsed}, remaining=${this.remainingTokens}, strict=${this.strict})`;
  }
}

// ── Exceptions ────────────────────────────────────────────────

/**
 * Thrown in strict mode when a single request exceeds the per-window budget.
 * Catch this to handle oversized requests explicitly rather than silently waiting.
 */
class BudgetExceeded extends Error {
  constructor(message) {
    super(message);
    this.name = "BudgetExceeded";
  }
}

// ── Sub-agent convenience ─────────────────────────────────────

/**
 * Pre-configured TokenThrottle for sub-agent use.
 *
 * Enforces a conservative shared-budget ceiling automatically so sub-agents
 * don't accidentally consume the full API key TPM and starve the main session.
 *
 * Default: caps effective TPM to 50% of the total limit (15K on a 30K key).
 *
 * @example
 *   const throttle = new SubAgentThrottle();            // 15K TPM ceiling
 *   const throttle = new SubAgentThrottle({ budgetFraction: 0.4 }); // 12K TPM
 */
class SubAgentThrottle extends TokenThrottle {
  /**
   * @param {Object} options
   * @param {number} [options.totalTpm=30000] - Full API key TPM limit (shared across all agents).
   * @param {number} [options.budgetFraction=0.5] - Fraction of totalTpm this sub-agent may use.
   * @param {...*} rest - Passed through to TokenThrottle (e.g. strict, chunkSize).
   */
  constructor({ totalTpm = 30000, budgetFraction = 0.5, ...rest } = {}) {
    super({
      tokensPerMinute: totalTpm,
      margin: budgetFraction,
      ...rest,
    });
  }

  toString() {
    return `SubAgentThrottle(tpm=${this.tpm}, budget=${this.budget}, used=${this._tokensUsed}, remaining=${this.remainingTokens}, strict=${this.strict})`;
  }
}

// ── Convenience functions ─────────────────────────────────────

const fs = require("fs");
const https = require("https");
const http = require("http");

/**
 * Estimate tokens in a file without reading it (uses file size).
 * @param {string} filePath
 * @param {number} [charsPerToken=4]
 * @returns {number}
 */
function estimateFileTokens(filePath, charsPerToken = 4) {
  const stat = fs.statSync(filePath);
  return Math.max(1, Math.floor(stat.size / charsPerToken));
}

/**
 * Estimate tokens from a URL via HEAD request Content-Length.
 * @param {string} url
 * @param {number} [charsPerToken=4]
 * @returns {Promise<number|null>}
 */
function estimateUrlTokens(url, charsPerToken = 4) {
  return new Promise((resolve) => {
    const mod = url.startsWith("https") ? https : http;
    const req = mod.request(url, { method: "HEAD", timeout: 10000 }, (res) => {
      const length = res.headers["content-length"];
      resolve(length ? Math.max(1, Math.floor(parseInt(length) / charsPerToken)) : null);
    });
    req.on("error", () => resolve(null));
    req.on("timeout", () => {
      req.destroy();
      resolve(null);
    });
    req.end();
  });
}

module.exports = { TokenThrottle, SubAgentThrottle, BudgetExceeded, estimateFileTokens, estimateUrlTokens };

File: python/token_throttle.py
"""
token_throttle - Simple token rate-limit management for LLM API calls.

Zero dependencies. Drop into any project.

Usage:
    from token_throttle import TokenThrottle

    throttle = TokenThrottle(tokens_per_minute=30000)

    # Auto-chunk and throttle large content
    for chunk in throttle.consume(large_text):
        response = call_your_api(chunk)

    # Or manually gate individual calls
    throttle.wait_if_needed(estimated_tokens)
    response = call_your_api(content)

Sub-agent usage (enforces shared budget automatically):
    from token_throttle import SubAgentThrottle

    throttle = SubAgentThrottle()  # caps to 15K TPM by default
    for chunk in throttle.consume(large_text):
        response = call_your_api(chunk)

Strict mode (raises BudgetExceeded instead of silently waiting):
    from token_throttle import TokenThrottle, BudgetExceeded

    throttle = TokenThrottle(budget=15000, strict=True)
    try:
        throttle.wait_if_needed(25000)
    except BudgetExceeded:
        # handle oversized request
        ...
"""

import time
import re
from typing import Iterator, Optional


class TokenThrottle:
    """Track token usage within a rolling minute window and throttle to stay under budget."""

    def __init__(
        self,
        tokens_per_minute: int = 30_000,
        margin: float = 0.85,
        budget: Optional[int] = None,
        chunk_size: Optional[int] = None,
        chars_per_token: float = 4.0,
        strict: bool = False,
    ):
        """
        Args:
            tokens_per_minute: Your TPM limit.
            margin: Safety factor (0-1). Effective budget = tpm * margin.
            budget: Hard cap on effective TPM (useful for shared-limit sub-agents).
                    Ceiling = min(budget, tpm * margin). Defaults to tpm * margin.
            chunk_size: Max tokens per yielded chunk. Defaults to effective budget.
            chars_per_token: Characters-per-token ratio for estimation (4 ≈ English).
            strict: If True, raise BudgetExceeded instead of waiting when a single
                    request exceeds the budget. Useful for catching oversized calls early.
        """
        self.tpm = tokens_per_minute
        self.margin = margin
        self.strict = strict
        _effective = int(tokens_per_minute * margin)
        self.budget = min(budget, _effective) if budget is not None else _effective
        self.chunk_size = chunk_size or self.budget
        self.chars_per_token = chars_per_token

        self._tokens_used = 0
        self._window_start = time.monotonic()

    # ── Estimation ──────────────────────────────────────────────

    def estimate(self, text: str) -> int:
        """Estimate token count from text length."""
        return max(1, int(len(text) / self.chars_per_token))

    # ── Window management ───────────────────────────────────────

    def _maybe_reset_window(self) -> None:
        elapsed = time.monotonic() - self._window_start
        if elapsed >= 60.0:
            self._tokens_used = 0
            self._window_start = time.monotonic()

    def reset(self) -> None:
        """Manually reset the usage window."""
        self._tokens_used = 0
        self._window_start = time.monotonic()

    # ── Throttling ──────────────────────────────────────────────

    def wait_if_needed(self, tokens: int) -> float:
        """
        Reserve `tokens` from the budget, sleeping if necessary.

        Returns seconds slept (0 if no wait was needed).

        Raises:
            BudgetExceeded: If strict=True and `tokens` exceeds the per-window budget.
        """
        if self.strict and tokens > self.budget:
            raise BudgetExceeded(
                f"Request of {tokens} tokens exceeds budget of {self.budget} TPM. "
                "Reduce the request size or lower strict=True."
            )

        self._maybe_reset_window()

        if self._tokens_used + tokens <= self.budget:
            self._tokens_used += tokens
            return 0.0

        # Need to wait for window reset
        elapsed = time.monotonic() - self._window_start
        sleep_time = max(0.0, 60.0 - elapsed + 0.5)  # +0.5s buffer
        if sleep_time > 0:
            time.sleep(sleep_time)

        # Reset after sleeping
        self._tokens_used = tokens
        self._window_start = time.monotonic()
        return sleep_time

    # ── Chunking ────────────────────────────────────────────────

    def _split_text(self, text: str, max_chars: int) -> list[str]:
        """Split text into chunks of roughly max_chars, breaking on paragraph/sentence/word boundaries."""
        if len(text) <= max_chars:
            return [text]

        chunks = []
        remaining = text

        while remaining:
            if len(remaining) <= max_chars:
                chunks.append(remaining)
                break

            # Try to break at paragraph boundary
            slice_ = remaining[:max_chars]
            break_pos = slice_.rfind("\n\n")

            # Fall back to sentence boundary
            if break_pos < max_chars // 4:
                match = None
                for m in re.finditer(r"[.!?]\s+", slice_):
                    match = m
                break_pos = match.end() if match and match.end() > max_chars // 4 else -1

            # Fall back to word boundary
            if break_pos < max_chars // 4:
                break_pos = slice_.rfind(" ")

            # Last resort: hard cut
            if break_pos < max_chars // 4:
                break_pos = max_chars

            chunks.append(remaining[:break_pos].rstrip())
            remaining = remaining[break_pos:].lstrip()

        return chunks

    # ── Main interface ──────────────────────────────────────────

    def consume(self, text: str) -> Iterator[str]:
        """
        Yield throttled chunks of text that fit within the token budget.

        Each chunk is guaranteed to be under `chunk_size` tokens.
        Automatically sleeps between chunks when the budget is exhausted.
        """
        max_chars = int(self.chunk_size * self.chars_per_token)
        chunks = self._split_text(text, max_chars)

        for chunk in chunks:
            tokens = self.estimate(chunk)
            self.wait_if_needed(tokens)
            yield chunk

    # ── Utilities ───────────────────────────────────────────────

    @property
    def remaining_tokens(self) -> int:
        """Tokens remaining in the current window."""
        self._maybe_reset_window()
        return max(0, self.budget - self._tokens_used)

    @property
    def seconds_until_reset(self) -> float:
        """Seconds until the current window resets."""
        elapsed = time.monotonic() - self._window_start
        return max(0.0, 60.0 - elapsed)

    def __repr__(self) -> str:
        return (
            f"TokenThrottle(tpm={self.tpm}, budget={self.budget}, "
            f"used={self._tokens_used}, remaining={self.remaining_tokens}, "
            f"strict={self.strict})"
        )


# ── Exceptions ─────────────────────────────────────────────────

class BudgetExceeded(Exception):
    """
    Raised in strict mode when a single request exceeds the per-window budget.

    Catch this to handle oversized requests explicitly rather than silently waiting.
    """
    pass


# ── Sub-agent convenience ───────────────────────────────────────

class SubAgentThrottle(TokenThrottle):
    """
    Pre-configured TokenThrottle for sub-agent use.

    Enforces a conservative shared-budget ceiling automatically so sub-agents
    don't accidentally consume the full API key TPM and starve the main session.

    Default behavior: caps effective TPM to 50% of the total limit (15K on a 30K key).

    Usage:
        throttle = SubAgentThrottle()            # 15K TPM ceiling
        throttle = SubAgentThrottle(budget_fraction=0.4)  # 12K TPM ceiling
    """

    def __init__(
        self,
        total_tpm: int = 30_000,
        budget_fraction: float = 0.5,
        **kwargs,
    ):
        """
        Args:
            total_tpm: The full API key TPM limit (shared across all agents).
            budget_fraction: Fraction of total_tpm this sub-agent may use (default 0.5 → 15K).
            **kwargs: Passed through to TokenThrottle (e.g. strict=True, chunk_size=...).
        """
        super().__init__(
            tokens_per_minute=total_tpm,
            margin=budget_fraction,
            **kwargs,
        )

    def __repr__(self) -> str:
        return (
            f"SubAgentThrottle(tpm={self.tpm}, budget={self.budget}, "
            f"used={self._tokens_used}, remaining={self.remaining_tokens}, "
            f"strict={self.strict})"
        )


# ── Convenience for quick scripts ──────────────────────────────

def estimate_file_tokens(path: str, chars_per_token: float = 4.0) -> int:
    """Estimate tokens in a file without reading it fully (uses file size)."""
    import os
    size = os.path.getsize(path)
    return max(1, int(size / chars_per_token))


def estimate_url_tokens(url: str, chars_per_token: float = 4.0, timeout: int = 10) -> Optional[int]:
    """Estimate tokens from a URL via HEAD request Content-Length. Returns None if unavailable."""
    import urllib.request
    try:
        req = urllib.request.Request(url, method="HEAD")
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            length = resp.headers.get("Content-Length")
            if length:
                return max(1, int(int(length) / chars_per_token))
    except Exception:
        pass
    return None
